File: examples/java/sdk-reference/error-handling/exception-hierarchy.java
import software.amazon.lambda.durable.exception.DurableExecutionException;
import software.amazon.lambda.durable.exception.DurableOperationException;
import software.amazon.lambda.durable.exception.UnrecoverableDurableExecutionException;
import software.amazon.lambda.durable.exception.StepException;
import software.amazon.lambda.durable.exception.StepFailedException;
import software.amazon.lambda.durable.exception.StepInterruptedException;
import software.amazon.lambda.durable.exception.CallbackException;
import software.amazon.lambda.durable.exception.CallbackFailedException;
import software.amazon.lambda.durable.exception.SerDesException;
import software.amazon.lambda.durable.exception.IllegalDurableOperationException;
import software.amazon.lambda.durable.exception.NonDeterministicExecutionException;

// DurableExecutionException
//   UnrecoverableDurableExecutionException — execution terminated immediately
//     IllegalDurableOperationException     — illegal SDK operation detected
//     NonDeterministicExecutionException   — non-deterministic replay detected
//   DurableOperationException              — operation-level failure with details
//     StepException
//       StepFailedException                — step failed after retries exhausted
//       StepInterruptedException           — at-most-once step interrupted
//     CallbackException
//       CallbackFailedException            — callback failed with error from external system
//   SerDesException                        — serialization or deserialization failed

File: examples/csharp/sdk-reference/error-handling/exception-hierarchy.cs
using Amazon.Lambda.DurableExecution;

// DurableExecutionException                — base class for all SDK exceptions
//   NonDeterministicExecutionException     — non-deterministic replay detected
//   StepException                          — step failed after retries exhausted
//     StepInterruptedException             — at-most-once step interrupted mid-attempt
//   ChildContextException                  — child context's user function failed
//   InvokeException                        — chained invoke failed
//     InvokeFailedException                — invoked function ran and threw (FAILED)
//     InvokeTimedOutException              — invoked function reached TIMED_OUT
//     InvokeStoppedException               — invocation stopped administratively
//   CallbackException                      — parent class for callback failures
//     CallbackFailedException              — external system reported a failure
//     CallbackTimeoutException             — callback (or heartbeat) timeout elapsed
//     CallbackSubmitterException           — submitter step failed after retries
//   ParallelException                      — parallel exceeded failure tolerance
//   MapException                           — map exceeded failure tolerance
//   WaitForConditionException              — wait-for-condition hit max attempts
//
// OperationCanceledException (System) — NOT a DurableExecutionException. Thrown
//   when the linked CancellationToken trips (workflow shutdown or cooperative
//   short-circuit). Treated as cancellation, not a step failure.
